File: php/features.php
<?php
declare(strict_types=1);

// Terra SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/DebugFeature.php';
require_once __DIR__ . '/feature/IdempotencyFeature.php';
require_once __DIR__ . '/feature/MetricsFeature.php';
require_once __DIR__ . '/feature/PagingFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class TerraFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new TerraBaseFeature();
            case "debug":
                return new TerraDebugFeature();
            case "idempotency":
                return new TerraIdempotencyFeature();
            case "metrics":
                return new TerraMetricsFeature();
            case "paging":
                return new TerraPagingFeature();
            case "ratelimit":
                return new TerraRatelimitFeature();
            case "retry":
                return new TerraRetryFeature();
            case "test":
                return new TerraTestFeature();
            case "timeout":
                return new TerraTimeoutFeature();
            default:
                return new TerraBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "debug":
            case "idempotency":
            case "metrics":
            case "paging":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}

File: py/terra_sdk/features.py
# Terra SDK feature factory

from terra_sdk.feature.base_feature import TerraBaseFeature
from terra_sdk.feature.debug_feature import TerraDebugFeature
from terra_sdk.feature.idempotency_feature import TerraIdempotencyFeature
from terra_sdk.feature.metrics_feature import TerraMetricsFeature
from terra_sdk.feature.paging_feature import TerraPagingFeature
from terra_sdk.feature.ratelimit_feature import TerraRatelimitFeature
from terra_sdk.feature.retry_feature import TerraRetryFeature
from terra_sdk.feature.test_feature import TerraTestFeature
from terra_sdk.feature.timeout_feature import TerraTimeoutFeature


_FEATURES = {
    "base": lambda: TerraBaseFeature(),
    "debug": lambda: TerraDebugFeature(),
    "idempotency": lambda: TerraIdempotencyFeature(),
    "metrics": lambda: TerraMetricsFeature(),
    "paging": lambda: TerraPagingFeature(),
    "ratelimit": lambda: TerraRatelimitFeature(),
    "retry": lambda: TerraRetryFeature(),
    "test": lambda: TerraTestFeature(),
    "timeout": lambda: TerraTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES
